exporter/exporterhelper/batch_sender.go

```diff
@@ -149,27 +155,39 @@ func (bs *batchSender) send(ctx context.Context, req Request) error {
 // sendMergeSplitBatch sends the request to the batch which may be split into multiple requests.
 func (bs *batchSender) sendMergeSplitBatch(ctx context.Context, req Request) error {
 	bs.mu.Lock()
-	bs.activeRequests.Add(1)
-	defer bs.activeRequests.Add(^uint64(0))
 
 	reqs, err := bs.mergeSplitFunc(ctx, bs.cfg.MaxSizeConfig, bs.activeBatch.request, req)
 	if err != nil || len(reqs) == 0 {
 		bs.mu.Unlock()
 		return err
 	}
+
+	bs.activeRequests.Add(1)
```

Comment:
I think this can be pushed under the following condition and we won't need the leftover

Reply:
It cannot be pushed down, because we still need the correct counts for `bs.activeRequests.Load() >= bs.concurrencyLimit`. However, I've refactored it to make it cleaner.

Reply:
My understanding of the concurrency control in the batch sender is to work with queue sender num_consumers and avoid meaningless waits. This means that if num_consumers=10 in queue sender, there will be at most 10 concurrent `send` calls issued to batch sender.

> If len(reqs) > 1 && bs.activeBatch.request == nil, we don't need to change requestsBlocked and flip bs.activeRequests back and forth.

I agree that requestsBlocked does not need to be changed as it bypasses activeBatch.

However, `bs.activeRequests` still needs to be flipped back and forth even in the case of `len(reqs) > 1 && bs.activeBatch.request == nil` because it blocks 1 consumer goroutine from queue sender.

Let's imagine a scenario where queue sender num_consumers=2.
- The first consumer C1 calls batch sender `send`, and hits the `len(reqs) > 1 && bs.activeBatch.request == nil` branch, and sends the requests (let's call them "R1") right away to the backend without touching activeBatch.
- If `bs.activeRequests` is NOT incremented while R1 are being sent, and a new consumer C2 calls batch sender `send` with `len(reqs) == 1`, when C2 hits `bs.isActiveBatchReady()`, the concurrency check `bs.activeRequests.Load() >= bs.concurrencyLimit` will be `1 >= 2` (where the 1 is from C2, because C1 is not accounted), then C2 will be subject to a meaningless wait for C1 to finish sending R1.
- If `bs.activeRequests` IS incremented while R1 are being sent, `bs.isActiveBatchReady()` will be true because of the concurrency check, and C2 will be sent immediately and will avoid the meaningless wait.

Reply:
The goal of this PR is to
1. correctly keep track of `bs.activeRequests`, which denotes the number of `send` waiting for next sender in chain to return. This is already done before this PR, but in a way that is vulnerable to unfavorable goroutine scheduling
2. decrease `bs.activeRequests` by the number of requests blocked by an activeBatch at once, so that it workarounds the "bug" mentioned in (1).

> If len(reqs) > 1 && bs.activeBatch.request == nil, we don't need to ... flip bs.activeRequests back and forth

If we do this, `bs.activeRequests` will be incorrect. It will be an underestimate of actual ongoing `send`.

Reply:
Ok, I get it now. Thanks!